File: packages/agent/src/events/types/streaming.ts
/**
 * @fileoverview Streaming Events
 *
 * Events for real-time streaming reconstruction.
 */

import type { BaseEvent } from './base.js';
import type { TokenUsage } from './token-usage.js';

// =============================================================================
// Normalized Token Usage
// =============================================================================

/**
 * Normalized token usage with semantic clarity for different UI components.
 * Handles the semantic differences in how different providers report tokens:
 * - Anthropic: inputTokens is NEW tokens only (excludes cache)
 * - OpenAI/Codex/Gemini: inputTokens is FULL context sent
 */
export interface NormalizedTokenUsage {
  /** Per-turn NEW input tokens (for stats line display) */
  newInputTokens: number;
  /** Output tokens for this turn */
  outputTokens: number;
  /** Total context window size (for progress pill) */
  contextWindowTokens: number;
  /** Raw input tokens as reported by provider (for billing/debugging) */
  rawInputTokens: number;
  /** Tokens read from cache (Anthropic/OpenAI) */
  cacheReadTokens: number;
  /** Tokens written to cache (Anthropic only) */
  cacheCreationTokens: number;
}

// =============================================================================
// Streaming Events
// =============================================================================

/**
 * Turn start event
 */
export interface StreamTurnStartEvent extends BaseEvent {
  type: 'stream.turn_start';
  payload: {
    turn: number;
  };
}

/**
 * Turn end event
 */
export interface StreamTurnEndEvent extends BaseEvent {
  type: 'stream.turn_end';
  payload: {
    turn: number;
    tokenUsage: TokenUsage;
    /**
     * Normalized token usage with semantic clarity for different UI components.
     * Handles provider semantic differences (Anthropic vs OpenAI/Codex/Gemini).
     */
    normalizedUsage?: NormalizedTokenUsage;
    /** Cost for this turn in USD */
    cost?: number;
  };
}

/**
 * Text delta for streaming reconstruction
 */
export interface StreamTextDeltaEvent extends BaseEvent {
  type: 'stream.text_delta';
  payload: {
    delta: string;
    turn: number;
    /** Content block index */
    blockIndex?: number;
  };
}

/**
 * Thinking delta for streaming reconstruction
 */
export interface StreamThinkingDeltaEvent extends BaseEvent {
  type: 'stream.thinking_delta';
  payload: {
    delta: string;
    turn: number;
  };
}

File: packages/agent-rs/crates/tron-events/src/types/payloads/streaming.rs
//! Streaming event payloads: turn start/end, text/thinking deltas.

use serde::{Deserialize, Serialize};

use super::token_usage::{TokenRecord, TokenUsage};

/// Payload for `stream.turn_start` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamTurnStartPayload {
    /// Turn number.
    pub turn: i64,
}

/// Payload for `stream.turn_end` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamTurnEndPayload {
    /// Turn number.
    pub turn: i64,
    /// Token usage for this turn.
    pub token_usage: TokenUsage,
    /// Canonical token record.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_record: Option<TokenRecord>,
    /// Cost in USD.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost: Option<f64>,
}

/// Payload for `stream.text_delta` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamTextDeltaPayload {
    /// Text fragment.
    pub delta: String,
    /// Turn number.
    pub turn: i64,
    /// Block index within the response.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub block_index: Option<i64>,
}

/// Payload for `stream.thinking_delta` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamThinkingDeltaPayload {
    /// Thinking text fragment.
    pub delta: String,
    /// Turn number.
    pub turn: i64,
}
